### PR TITLE
Update callbacks.py

### LINKED ISSUE #2849: Improve Documentation for ValidationCallback to Assist with Hyperparameter Tuning
## 🚀 Feature

I took a closer look at ValidationCallback in deepchem/models/callbacks.py and I don't think any more code needs to be written. There is already an attribute named self._best_score and an argument named save_on_minimum which allows choosing minimize or maximize. 

I have a few ideas: 1) Add a function to ValidationCallback called GetBestScore() or something. It may honestly not be necessary as like I said, self._best_score is already there. 2) Edit the dc.hyper.GridHyperparamOpt tuner to use the best validation score. I think it currently evaluates the FINAL model on the validation set instead of the BEST model but please correct me if I'm wrong. Then document this change clearly so others who wish to create a new hyperparameter optimizer will have an example to follow.

I think as a follow up, actual early stopping where the model saves the best model and stops if it doesn't improve after n steps would be a really nice feature to have, although I'm sure it'd be very complicated to implement across so many models. 

## Motivation

Allowing users to access the best validation score for the purpose of early stopping is very important for hyperparameter optimization. This feature is already there but the question is how do we make it easier to use/develop with. 

## Additional context

### PR DESCRIPTION
**Pull request by TheRealSalmon**
Update callbacks to always update best score and to implement a getter.

# Pull Request Template

## Description

Fix #2849

Changed the __call__ function to always update self._best_score. Previously it only updated it when a save_dir was specified. Also I added a getter for self._best_score.

## Type of change

Please check the option that is related to your PR.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
  - In this case, we recommend to discuss your modification on GitHub issues before creating the PR
- [ ] Documentations (modification for documents)

## Checklist

- [x] My code follows [the style guidelines of this project](https://deepchem.readthedocs.io/en/latest/development_guide/coding.html)
  - [ ] Run `yapf -i <modified file>` and check no errors (**yapf version must be  0.22.0**)
  - [ ] Run `mypy -p deepchem` and check no errors
  - [ ] Run `flake8 <modified file> --count` and check no errors
  - [ ] Run `python -m doctest <modified file>` and check no errors
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: deepchem/models/callbacks.py [hunk 1]
```diff
@@ -86,14 +91,16 @@ def __call__(self, model, step):
...
-        self._best_score = score
+    if self.save_dir is not None and (self._best_score is None or
+                                      score < self._best_score):
+      model.save_checkpoint(model_dir=self.save_dir)
     if model.wandb_logger is not None:
       # Log data to Wandb
       data = {'eval/' + k: v for k, v in scores.items()}
       model.wandb_logger.log_data(data, step, dataset_id=id(self.dataset))
+
+  def get_best_score(self):
```
Comment (rbharath) [inline deepchem/models/callbacks.py:102]: Should be `_get_best_score`

File: deepchem/models/callbacks.py [hunk 2]
```diff
@@ -78,6 +78,11 @@ def __call__(self, model, step):
     if step % self.interval != 0:
       return
     scores = model.evaluate(self.dataset, self.metrics, self.transformers)
+    score = scores[self.metrics[self.save_metric].name]
```
Comment (rbharath) [inline deepchem/models/callbacks.py:81]: We should add a unit test for the new functionality

File: deepchem/models/callbacks.py (hunk 1 above)
Comment (peastman) [reply deepchem/models/callbacks.py:102]: I think it's intended to be a public method, so the name is fine.  It just needs a docstring.

**Comment by peastman** [comment]
Did you intend to close this?
